### PR TITLE
[ARCTIC-925] Introduce SpillableMap to avoid OOM for optimizing and merge-on-read

### PR DESCRIPTION
**Pull request by majin1102**
## Why are the changes needed?
 
Intruduce #925 data structures, while not used yet.

## Brief change log

  1. Introduce StructLikeSpillableMap with RocksDB backend for optimizing and merge-on-read
  2. Refactor current StructLikeMap to StructLikeMemoryMap and organize related classes into a utils.map packages
  3. For current optimizing and merge-on-read, still use StructLikeMemoryMap. AMS will determine whether using StructLikeSpillableMap or StructLikeMemoryMap in other PRs

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes )
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/data/DefaultKeyedFile.java [hunk 1]
```diff
@@ -41,13 +41,31 @@ public class DefaultKeyedFile implements PrimaryKeyedFile, Serializable {
...
+      }
+    }
+    if (internalFile.upperBounds() != null) {
+      ByteBuffer maxOffsetBuffer = internalFile.upperBounds().get(MetadataColumns.FILE_OFFSET_FILED_ID);
+      if (maxOffsetBuffer != null) {
+        maxLsn = ChangedLsn.of(meta.transactionId(), Conversions.fromByteBuffer(Types.LongType.get(), maxOffsetBuffer));
+      } else {
+        maxLsn = ChangedLsn.of(meta.transactionId(), Long.MAX_VALUE);
+      }
+    }
```
Comment (YesOrNo828) [inline core/src/main/java/com/netease/arctic/data/DefaultKeyedFile.java:68]: Maybe we could extract the common logic into a private method, for example:

```
private ChangedLsn parseChangedLsn(Map<Integer, ByteBuffer> map, FileMeta meta) {
    ByteBuffer offsetBuffer = map.get(MetadataColumns.FILE_OFFSET_FILED_ID);
      if (maxOffsetBuffer != null) {
        lsn = ChangedLsn.of(meta.transactionId(), Conversions.fromByteBuffer(Types.LongType.get(), offsetBuffer));
      } else {
        lsn = ChangedLsn.of(meta.transactionId(), Long.MAX_VALUE);
      }
      return lsn;
}
```

**Review by YesOrNo828 — COMMENTED** [review]
@majin1102  I left some comments. 
BTW, you should add the apache license at the head of each new file.

The following steps may help you automate in IntelliJ by adding a Copyright profile:

1、Go to “Settings” → “Editor” → “Copyright” → “Copyright Profiles”.

2、Add a new profile and name it “Apache”.

3、Add the following text as the license text:

4、Go to “Editor” → “Copyright” and choose the “Apache” profile as the default profile for this project.

5、Click “Apply”.

File: core/src/main/java/com/netease/arctic/utils/LocalFileUtils.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.netease.arctic.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Comparator;
+
+public class LocalFileUtils {
```
[the remark this answers is not in the document]
Comment (majin1102) [reply core/src/main/java/com/netease/arctic/utils/LocalFileUtils.java:8]: Local is used for local file system, and Table is for distributed, or it is used by table operations

File: core/src/main/java/com/netease/arctic/data/DefaultKeyedFile.java (hunk 1 above)
Comment (majin1102) [reply core/src/main/java/com/netease/arctic/data/DefaultKeyedFile.java:68]: this is original code, you could arise this in other PRs

File: core/src/main/java/com/netease/arctic/utils/map/RocksDBBackend.java [hunk 3]
```diff
@@ -0,0 +1,359 @@
+package com.netease.arctic.utils.map;
```
[the remark this answers is not in the document]
Comment (majin1102) [reply core/src/main/java/com/netease/arctic/utils/map/RocksDBBackend.java:19]: > @majin1102 I left some comments. BTW, you should add the apache license at the head of each new file.
> 
> The following steps may help you automate in IntelliJ by adding a Copyright profile:
> 
> 1、Go to “Settings” → “Editor” → “Copyright” → “Copyright Profiles”.
> 
> 2、Add a new profile and name it “Apache”.
> 
> 3、Add the following text as the license text:
> 
> 4、Go to “Editor” → “Copyright” and choose the “Apache” profile as the default profile for this project.
> 
> 5、Click “Apply”.

These have been done

**Review by YesOrNo828 — APPROVED** [review]

**Pull request merged**